Closes #4067: Fennec logins migration

grigoryk opened the pull request:
---
<!-- Text above this line will be added to the commit once "bors" merges this PR -->

### Pull Request checklist
<!-- Before submitting the PR, please address each item -->
- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### After merge
- [ ] **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- [ ] **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

grigoryk commented:
This PR doesn't expose `master password` as part of FennecMigrator's public API - but otherwise, it supports decryption of logins dbs protected with MP.

grigoryk commented:
Fenix PR that enables logins migration: https://github.com/mozilla-mobile/fenix/pull/6930

csadilek commented on components/support/migration/src/main/java/mozilla/components/support/migration/FennecLoginsMigration.kt:
```diff
@@ -0,0 +1,541 @@
+            it.moveToNext()
+
+            val globalSalt = it.getBlob(it.getColumnIndexOrThrow("item1"))
+            val item2 = it.getBlob(it.getColumnIndexOrThrow("item2"))
```
Can we add the details from your comment https://github.com/mozilla-mobile/android-components/issues/4067#issuecomment-558424220 to describe what item2 is?

csadilek commented on components/support/migration/src/main/java/mozilla/components/support/migration/FennecLoginsMigration.kt:
```diff
@@ -0,0 +1,541 @@
+                    )
+                    records.add(fullRecord)
+                } catch (e: Exception) {
+                    crashReporter.submitCaughtException(e)
```
Here we'll send a caught exception to Sentry/Socorro for every incorrect record. How likely are incorrect records? Should we combine into one exception instead?

csadilek replied:
Actually, yes, I think we should move this out of the loop. `decrypt` could throw a `NoSuchAlgorithmException` and then we'd be logging this a lot...

csadilek commented on components/support/migration/src/main/java/mozilla/components/support/migration/FennecLoginsMigration.kt:
```diff
@@ -0,0 +1,541 @@
+                    return@use it.getBlob(it.getColumnIndexOrThrow("a11"))
+                }
+            }
+            null
```
nit: `return@use` here as well to make it easier to read?

csadilek commented on components/support/migration/src/main/java/mozilla/components/support/migration/FennecLoginsMigration.kt:
```diff
@@ -0,0 +1,541 @@
+    /**
+     * Decrypts [ciphertext] using TDES.
+     */
+    private fun decryptTDes(key: ByteArray, iv: ByteArray, ciphertext: ByteArray): ByteArray {
```
nit `descryptTDES`?

grigoryk commented:
@csadilek addresses all of your feedback.

I've reached out to @mozkeeler for feedback, as well. Dana thinks this is basically fine (non-ideal as there are hand-rolled ASN.1 decoder bits, but alas). Main concern is that we should be careful about running this migration while Gecko is around. "NSS has historically had issues when anything else touches cert9.db/key4.db".
But, It will _probably_ be fine even if we overlap with NSS, somehow - "since it's read-only, the worst that should happen is some slowness".

grigoryk commented:
bors r=csadilek
